core/unit_test/TestPolicyConstruction.hpp

```diff
@@ -691,29 +691,34 @@ void check_semiregular() {
   static_assert(std::is_destructible<Policy>::value, "");
 }
 
+template <template <class...> class Policy, class... Args>
+void check_conversions() {
+  struct WorkTag {};
+  static_assert(
+      std::is_convertible<Policy<Args...>, Policy<Args..., WorkTag>>::value,
+      "");
+  static_assert(
+      std::is_convertible<Policy<Args..., WorkTag>, Policy<Args...>>::value,
+      "");
+  using IndexType = Kokkos::IndexType<long long int>;
+  static_assert(
+      std::is_convertible<Policy<Args...>, Policy<Args..., IndexType>>::value,
+      "");
+  static_assert(
+      std::is_convertible<Policy<Args..., IndexType>, Policy<Args...>>::value,
+      "");
+}
+
 TEST(TEST_CATEGORY, policy_construction) {
   check_semiregular<Kokkos::RangePolicy<TEST_EXECSPACE>>();
   check_semiregular<Kokkos::TeamPolicy<TEST_EXECSPACE>>();
 
+  check_conversions<Kokkos::RangePolicy, TEST_EXECSPACE>();
+  check_conversions<Kokkos::TeamPolicy, TEST_EXECSPACE>();
+  check_conversions<Kokkos::MDRangePolicy, TEST_EXECSPACE, Kokkos::Rank<2>>();
+
   TestRangePolicyConstruction<TEST_EXECSPACE>();
   TestTeamPolicyConstruction<TEST_EXECSPACE>();
 }
 
-template <template <class...> class Policy, class... Args>
-void check_converting_constructor_add_work_tag(Policy<Args...> const& policy) {
-  // Not the greatest but at least checking it compiles
-  struct WorkTag {};
-  Policy<Args..., WorkTag> policy_with_tag = policy;
```

Comment:
After talking to Christian, I realized the issue is we were copying uninitialized memory.
Default constructors for `RangePolicy` and `TeamPolicy` do not initialize all their members.

Reply:
Turns out it was UB but not from the converting constructor.  It was in `TeamPolicyInternal<Threads,...> default constructor.  c.f. #3512

Reply:
I still would like to see that assignment actually happen here. Sure actually I would like to see an assignment and then a comparison of intenral arguments but its not clear that that can even be made possible (like the additional WorkTag should likely make the comparison unequal. Otherwise ok. Or we can just close this thing since its now working and the test isn't really much better than before.